primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Implementation of a Namespaced Merkle Tree.
+use alloc::vec;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::vec::Vec;
+use core::{fmt::Debug, hash::Hash, marker::PhantomData};
+
+use crate::errors::PrimitivesError;
+
+use super::{AppendableMerkleTreeScheme, DigestAlgorithm, Element, Index, NodeValue};
+
+/// Namespaced Merkle Tree where leaves are sorted by a namespace identifier.
+pub trait NamespacedMerkleTreeScheme: AppendableMerkleTreeScheme
+where
+    Self::Element: Namespaced<Self::NamespaceId>,
+{
+    /// Namespace proof type
+    type NamespaceProof: Clone;
+    /// Namespace type
+    type NamespaceId: Namespace;
+
+    /// Returns the entire set of leaves corresponding to a given namespace and
+    /// a completeness proof
+    fn get_namespace_leaves_and_proof(
+        &self,
+        namespace: Self::NamespaceId,
+    ) -> (Vec<Self::Element>, Self::NamespaceProof);
+
+    /// Verifies the completeness proof for a given set of leaves and a
+    /// namespace
+    fn verify_namespace_proof(
+        leaves: &[Self::Element],
+        proof: Self::NamespaceProof,
+        namespace: Self::NamespaceId,
+    ) -> Result<(), PrimitivesError>;
+}
+
+/// NamespacedHasher wraps a standard hash function (implementer of
+/// DigestAlgorithm), turning it into a hash function that tags internal nodes
+/// with namespace ranges
+pub struct NamespacedHasher<H, E, I, T, N>
+where
+    H: DigestAlgorithm<E, I, T>,
+    E: Element + Namespaced<N>,
+    N: Namespace,
+    I: Index,
+    T: NodeValue,
+{
+    phantom1: PhantomData<H>,
+    phantom2: PhantomData<E>,
+    phantom3: PhantomData<I>,
+    phantom4: PhantomData<T>,
+    phantom5: PhantomData<N>,
+}
+
+/// Trait indicating that a leaf has a namespace
+pub trait Namespaced<N: Namespace> {
+    /// Returns the namespace of the leaf
+    fn get_namespace(&self) -> N;
+}
+
+/// Trait indicating that a digest algorithm can bind namespaces
+pub trait BindNamespace<E, I, T, N>: DigestAlgorithm<E, I, T>
```

Comment:
Not sure this is the best name - we need this trait so that we can generate a hash that binds an internal hash to its namespace range

Reply:
The name does not looks bad to me.